Centaurus.DAL/Mongo/MongoStorage.cs

```diff
@@ -260,89 +243,62 @@ public async Task<int> GetFirstPriceHistoryFrameDate(int market, PriceHistoryPer
-            while (true)
+            var quanta = update.Quanta;
+
+            //var retries = 1;
+            //var maxTries = 5;
+            //var isCommitInvoked = false;
+            //while (true)
+            //{
+            using (var session = await client.StartSessionAsync())
             {
-                using (var session = await client.StartSessionAsync())
+                var result = await session.WithTransactionAsync<bool>(async (s, ct) =>
                 {
-                    try
-                    {
-                        session.StartTransaction();
-                        try
-                        {
-                            var updateTasks = new List<Task>();
-
-                            if (stellarUpdates != null)
-                                updateTasks.Add(constellationStateCollection.BulkWriteAsync(session, new WriteModel<ConstellationState>[] { stellarUpdates }));
+                    //try
+                    //{
+                    var updateTasks = new List<Task>();
 
-                            if (update.ConstellationSettings != null)
-                                updateTasks.Add(settingsCollection.InsertOneAsync(session, update.ConstellationSettings));
+                    if (stellarUpdates != null)
+                        updateTasks.Add(constellationStateCollection.BulkWriteAsync(s, new WriteModel<ConstellationState>[] { stellarUpdates }));
 
-                            if (update.Assets != null && update.Assets.Count > 0)
-                                updateTasks.Add(assetsCollection.InsertManyAsync(session, update.Assets));
+                    if (update.ConstellationSettings != null)
+                        updateTasks.Add(settingsCollection.InsertOneAsync(s, update.ConstellationSettings));
 
-                            if (accountUpdates != null)
-                                updateTasks.Add(accountsCollection.BulkWriteAsync(session, accountUpdates));
+                    if (update.Assets != null && update.Assets.Count > 0)
+                        updateTasks.Add(assetsCollection.InsertManyAsync(s, update.Assets));
 
-                            if (balanceUpdates != null)
-                                updateTasks.Add(balancesCollection.BulkWriteAsync(session, balanceUpdates));
+                    if (accountUpdates != null)
+                        updateTasks.Add(accountsCollection.BulkWriteAsync(s, accountUpdates));
 
-                            if (orderUpdates != null)
-                                updateTasks.Add(ordersCollection.BulkWriteAsync(session, orderUpdates));
+                    if (balanceUpdates != null)
+                        updateTasks.Add(balancesCollection.BulkWriteAsync(s, balanceUpdates));
 
-                            SaveQuanta(ref updateTasks, quanta, session);
+                    if (orderUpdates != null)
+                        updateTasks.Add(ordersCollection.BulkWriteAsync(s, orderUpdates));
```

Comment:
I thought about this for a while. Those current state updates might be the real reason why the db periodically chokes on updates. Every order, balance, and account update requires an object lookup before the actual update. If we save orders and balances in arrays within a parent account, it saves us 2 lookups, 2 random access data writes on every trade operation. Also, there are fewer index updates in this scenario. 
I'm aware that this will require initialization code rewrite, but I still think this refactoring is totally worth it. Could you please put `orders` and `balances` as subarrays in the `account` model? And you can cleanup unneeded fields, like ids and account reference from orders and balances. Let's do it in this iteration.